subsys/sched_server/cbs/cbs.c

```diff
@@ -0,0 +1,333 @@
+	 */
+	if (deadline > arrival) {
+		cbs_cycle_t budget = (cbs->budget.current << cbs->left_shift);
+		if (budget < (deadline - arrival) * cbs->bandwidth) {
+			return;
+		}
+	}
+	k_sched_lock();
+	cbs_replenish_due_to_condition(cbs, arrival);
+	k_thread_deadline_set(cbs->thread, (int)cbs->period);
+	k_sched_unlock();
+}
+
+static void cbs_budget_timer_start(cbs_t *cbs)
+{
+	if (cbs->start_cycle > 0) {
+		return;
+	}
+	cbs->start_cycle = cbs_get_now();
+	k_timer_start(cbs->timer, K_CYC((uint32_t)cbs->budget.current), K_NO_WAIT);
+}
+
+static void cbs_budget_timer_stop(cbs_t *cbs)
+{
+	if (!cbs || cbs->timer->status) {
+		return;
+	}
+	k_timer_stop(cbs->timer);
+}
+
+static inline void cbs_find_highest_shift_for(cbs_t *cbs)
+{
+	/*
+	 * This for loop finds the highest left-shift value
+	 * we can apply to the budget before overflowing it.
+	 *
+	 * The actual left shift logical is done because the
+	 * CBS condition checks an equation that, in paper,
+	 * uses decimal numbers within it. However, all math
+	 * is made here keeps the unsigned integer types of the
+	 * variables involved for performance reasons. So
+	 * the shift is applied in both sides of the equation
+	 * in an attempt to preserve the value resolution.
+	 *
+	 * more details in cbs_budget_restore_if_condition().
+	 */
+	for (unsigned int right_shift = 1; right_shift < 32; right_shift++) {
+		if ((INT_MAX >> right_shift) < cbs->budget.max) {
+			cbs->left_shift = right_shift - 1;
+			cbs->bandwidth = (cbs->budget.max << cbs->left_shift) / cbs->period;
+			break;
+		}
+	}
+}
+
+void cbs_thread(void *server_name, void *cbs_struct, void *cbs_args)
+{
+	cbs_job_t job;
+	cbs_t *cbs = (cbs_t *)cbs_struct;
+	cbs_arg_t *args = (cbs_arg_t *)cbs_args;
+
+	cbs->thread = k_current_get();
+	cbs->thread->cbs = cbs;
+	cbs->start_cycle = 0;
+	cbs->period = CBS_TICKS_TO_CYC(args->period.ticks);
+	cbs->budget.max = CBS_TICKS_TO_CYC(args->budget.ticks);
+	cbs->budget.current = cbs->budget.max;
+	cbs->left_shift = 0;
+	cbs->abs_deadline = 0;
+	cbs_find_highest_shift_for(cbs);
+
+	k_timer_init(cbs->timer, cbs_budget_timer_expired_callback, cbs_budget_timer_stop_callback);
+	k_timer_user_data_set(cbs->timer, cbs);
+
+#ifdef CONFIG_CBS_LOG
+	strncpy(cbs->name, (char *)server_name, CONFIG_CBS_THREAD_MAX_NAME_LEN - 1);
+#endif
+
+	cbs->is_idle = true;
+	cbs->is_initialized = true;
+
+	for (;;) {
+		/*
+		 * The CBS thread remains dormant while
+		 * there are no jobs to execute. Once they
+		 * arrive, timer starts, execution is made,
+		 * timer ends, budget is updated and
+		 * the cycle repeats.
+		 */
+		k_msgq_get(cbs->queue, &job, K_FOREVER);
+		cbs_budget_timer_start(cbs);
+		job.function(job.arg);
+		cbs_budget_timer_stop(cbs);
+
+#if defined(CONFIG_CBS_LOG) && defined(CONFIG_CBS_LOG_JOB_COMPLETE)
+		cbs_log(CBS_COMPLETED_JOB, cbs);
+#endif
+		cbs->is_idle = (cbs->queue->used_msgs == 0);
+	}
+}
+
+void cbs_thread_switched_in(struct k_thread *thread)
+{
+	if (!thread || !thread->cbs) {
```

Comment:
Why check thread for null here?  That would be a bug in the context switch primitives you hooked.  This is **extremely** performance-sensitive code.

Reply:
Sorry, it's just me used to include defensive code by default. I have removed that.